File: subprojects/xcfa/xcfa-analysis/src/main/java/hu/bme/mit/theta/xcfa/analysis/XcfaAction.java
/*
 *  Copyright 2017 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.xcfa.analysis;

import hu.bme.mit.theta.analysis.expr.StmtAction;
import hu.bme.mit.theta.common.Utils;
import hu.bme.mit.theta.core.stmt.Stmt;
import hu.bme.mit.theta.xcfa.model.XcfaEdge;
import hu.bme.mit.theta.xcfa.model.XcfaLocation;

import java.util.Collections;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

public final class XcfaAction extends StmtAction {

	private final XcfaEdge edge;
	private final List<Stmt> stmts;
	private final XcfaLocation source;
	private final XcfaLocation target;

	private XcfaAction(final XcfaLocation source, final XcfaLocation target, final XcfaEdge edge) {
		this.source = checkNotNull(source);
		this.target = checkNotNull(target);
		this.edge = checkNotNull(edge);
		this.stmts = Collections.unmodifiableList(edge.getStmts());
	}

	public static XcfaAction create(final XcfaEdge edge) {
		return new XcfaAction(edge.getSource(), edge.getTarget(), edge);
	}

	public XcfaLocation getSource() {
		return source;
	}

	public XcfaLocation getTarget() {
		return target;
	}

	@Override
	public List<Stmt> getStmts() {
		return stmts;
	}

	public XcfaEdge getEdge() {
		return edge;
	}

	@Override
	public String toString() {
		return Utils.lispStringBuilder(getClass().getSimpleName()).body().addAll(stmts).toString();
	}
}

File: subprojects/xcfa/xcfa-analysis/src/main/java/hu/bme/mit/theta/xcfa/analysis/XcfaAction.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.xcfa.analysis

import hu.bme.mit.theta.analysis.algorithm.bounded.ReversibleAction
import hu.bme.mit.theta.analysis.ptr.PtrAction
import hu.bme.mit.theta.analysis.ptr.WriteTriples
import hu.bme.mit.theta.core.stmt.Stmt
import hu.bme.mit.theta.xcfa.model.*
import hu.bme.mit.theta.xcfa.passes.flatten

data class XcfaAction
@JvmOverloads
constructor(
  val pid: Int,
  val edge: XcfaEdge,
  private val lastWrites: WriteTriples = emptyMap(),
  private val nextCnt: Int = 0,
) : PtrAction(lastWrites, nextCnt), ReversibleAction {

  val source: XcfaLocation = edge.source
  val target: XcfaLocation = edge.target
  val label: XcfaLabel = edge.label
  private val stmts: List<Stmt> = label.toStmt().flatten()

  constructor(
    pid: Int,
    source: XcfaLocation,
    target: XcfaLocation,
    label: XcfaLabel = NopLabel,
    metaData: MetaData = EmptyMetaData,
    lastWrites: WriteTriples = emptyMap(),
    nextCnt: Int = 0,
  ) : this(pid, XcfaEdge(source, target, label, metaData), lastWrites, nextCnt)

  override val stmtList: List<Stmt>
    get() = stmts

  override fun reverse(): ReversibleAction {
    return XcfaAction(pid, XcfaEdge(target, source, label, edge.metadata))
  }

  override fun toString(): String {
    return "$pid: $source -> $target [${getStmts()}]"
  }

  fun withLabel(sequenceLabel: SequenceLabel): XcfaAction {
    return XcfaAction(pid, source, target, sequenceLabel, edge.metadata, nextCnt = nextCnt)
  }

  fun withLastWrites(writeTriples: WriteTriples, nextCnt: Int): XcfaAction {
    return XcfaAction(pid, source, target, label, edge.metadata, writeTriples, nextCnt)
  }
}
